flink-runtime/src/main/java/org/apache/flink/runtime/rpc/taskexecutor/TaskExecutor.java

```diff
@@ -35,27 +79,634 @@
+						fraction, directMemorySize >> 20);
+				} else {
+					LOG.info("Limiting managed memory to {} of the maximum memory size ({} MB)," +
+						" memory will be allocated lazily.", fraction, directMemorySize >> 20);
+				}
+				memorySize = directMemorySize;
+			} else {
+				throw new RuntimeException("No supported memory type detected.");
+			}
+		}
+
+		// now start the memory manager
+		MemoryManager memoryManager;
+		try {
+			memoryManager = new MemoryManager(
+				memorySize,
+				taskManagerConfig.numberOfSlots(),
+				netConfig.networkBufferSize(),
+				memType,
+				preAllocateMemory);
+		} catch (OutOfMemoryError e) {
+			if (memType == MemoryType.HEAP) {
+				throw new Exception("OutOfMemory error (" + e.getMessage() +
+					") while allocating the TaskManager heap memory (" + memorySize + " bytes).", e);
+			} else if (memType == MemoryType.OFF_HEAP) {
+				throw new Exception("OutOfMemory error (" + e.getMessage() +
+					") while allocating the TaskManager off-heap memory (" + memorySize +
+					" bytes).Try increasing the maximum direct memory (-XX:MaxDirectMemorySize)", e);
+			} else {
+				throw e;
+			}
+		}
+
+		// start the I/O manager, it will create some temp directories.
+		IOManager ioManager = new IOManagerAsync(taskManagerConfig.tmpDirPaths());
+
+		if (leaderRetrievalService == null){
+			leaderRetrievalService = LeaderRetrievalUtils.createLeaderRetrievalService(configuration);
+		}
+
+		// create the actor properties (which define the actor constructor parameters)
+		Props tmProps = Props.create(
+			taskManagerClass,
+			taskManagerConfig,
+			resourceID,
+			connectionInfo,
+			memoryManager,
+			ioManager,
+			network,
+			taskManagerConfig.numberOfSlots(),
+			leaderRetrievalService);
+
+		ActorRef taskManagerActorRef;
+		if (taskManagerActorName != null && !taskManagerActorName.equals("")) {
+			taskManagerActorRef = actorSystem.actorOf(tmProps, taskManagerActorName);
+		} else {
+			taskManagerActorRef = actorSystem.actorOf(tmProps);
+		}
+
+		return taskManagerActorRef;
+	}
+
+	// --------------------------------------------------------------------------
+	//  Parsing and checking the TaskManager Configuration
+	// --------------------------------------------------------------------------
+
+	/**
+	 * Utility method to extract TaskManager config parameters from the configuration and to
+	 * sanity check them.
+	 *
+	 * @param configuration                 The configuration.
+	 * @param taskManagerHostname           The host name under which the TaskManager communicates.
+	 * @param localTaskManagerCommunication True, to skip initializing the network stack.
+	 *                                      Use only in cases where only one task manager runs.
+	 * @return A tuple (TaskManagerConfiguration, network configuration,
+	 * InstanceConnectionInfo, JobManager actor Akka URL).
+	 */
+	private static Tuple4<TaskManagerConfiguration, NetworkEnvironmentConfiguration, InstanceConnectionInfo, MemoryType>
```

Comment:
I would create a static class to hold the results instead of using Scala tuples.